### PR TITLE
Consistently use the repository's preferred package manager

### LINKED ISSUE #432: changeset publish does not replace workspace:* with current version in workspace before publishing
## Affected Packages

@changesets/apply-release-plan

and maybe others...

## Problem

@zkochan has already raised an issue with support of workspace ranges in dependencies, and he already proposed a solution for his immediate need of replacing workspace:x.y.z with x.y.z and it's working. However the case of "workspace:*" is not taken care of yet. The documented behavior in all package managers is for it to be replaced with the current version in the workspace.

changeset publish does not do that resulting in the NPM package being published with 'workspace:*' which obviously does not work.

## Proposed solution

Not sure, but @zkochan already had a comment regarding this at: https://github.com/atlassian/changesets/pull/311/files#r407058482 and to my knowledge yarn2 properly publishes such packages when invoked with 'yarn npm publish'. Not sure about pnpm.
Maybe the solution is to let the package manager handle this?

**Workaround so far:**
I am using yarn v2, so the workaround for me is ugly but produces consistent results. 
- Just stop at changeset version
- Use yarn npm publish to publish packages
- manually create git tags
- push to git

### LINKED ISSUE #598: [Feature Request] Support for Yarn 2 "workspace ranges"
## Affected Packages

`@changesets/cli`

## Problem

First off, thanks for the great tool!

Yarn 2 introduces a feature called [workspace ranges](https://yarnpkg.com/features/workspaces#workspace-ranges-workspace).

With workspace ranges, you can specify the intent of "depend on the current version of local workspace package, whatever the actual published version it may be". If you think about the workflow you're simutaneously modifying multiple packages, this can be extremely useful.

The transformation of `workspace:...` version to actual versions that npm can understand is handled by yarn at publish time.

> When a workspace is packed into an archive (whether it's through `yarn pack` or one of the publish commands like `yarn npm publish`), we dynamically replace any workspace: dependency by:
> 
> * The corresponding version in the target workspace (if you use * or a project-relative path)
> * The associated semver range (for any other range type)

However, [as current `changeset publish` command only supports either npm or pnpm as publish client](https://github.com/atlassian/changesets/blob/c426035565cfac518238c8bf32f3c496c66c0657/packages/cli/src/commands/publish/npm-utils.ts#L36-L56), this transformation cannot take place, and `changeset publish` aborts with:

```
🦋  error TypeError: Invalid comparator: packages/my-package
🦋  error     at Comparator.parse (/home/runner/work/npm-packages/npm-packages/.yarn/cache/semver-npm-5.7.1-40bcea106b-06ff0ed753.zip/node_modules/semver/semver.js:754:11)
🦋  error     at new Comparator (/home/runner/work/npm-packages/npm-packages/.yarn/cache/semver-npm-5.7.1-40bcea106b-06ff0ed753.zip/node_modules/semver/semver.js:737:8)
🦋  error     at Range.<anonymous> (/home/runner/work/npm-packages/npm-packages/.yarn/cache/semver-npm-5.7.1-40bcea106b-06ff0ed753.zip/node_modules/semver/semver.js:925:12)
🦋  error     at Array.map (<anonymous>)
🦋  error     at Range.parseRange (/home/runner/work/npm-packages/npm-packages/.yarn/cache/semver-npm-5.7.1-40bcea106b-06ff0ed753.zip/node_modules/semver/semver.js:924:13)
🦋  error     at Range.<anonymous> (/home/runner/work/npm-packages/npm-packages/.yarn/cache/semver-npm-5.7.1-40bcea106b-06ff0ed753.zip/node_modules/semver/semver.js:867:17)
🦋  error     at Array.map (<anonymous>)
🦋  error     at new Range (/home/runner/work/npm-packages/npm-packages/.yarn/cache/semver-npm-5.7.1-40bcea106b-06ff0ed753.zip/node_modules/semver/semver.js:866:40)
🦋  error     at versionPackage (/home/runner/work/npm-packages/npm-packages/.yarn/cache/@changesets-apply-release-plan-npm-5.0.0-19ff54c5d1-d43561e831.zip/node_modules/@changesets/apply-release-plan/dist/apply-release-plan.cjs.dev.js:160:9)
🦋  error     at /home/runner/work/npm-packages/npm-packages/.yarn/cache/@changesets-apply-release-plan-npm-5.0.0-19ff54c5d1-d43561e831.zip/node_modules/@changesets/apply-release-plan/dist/apply-release-plan.cjs.dev.js:312:12
```

## Proposed solution

If either

- `@changesets/cli` allows the configuration of publish command (Related issue: #580) or,
- it supports Yarn 2 out of the box

It would be awesome!

### LINKED ISSUE #813: Yarn not used for publishing
## Affected Packages
N/A

## Problem

Changesets doesn't seem to use yarn for publishing (only npm or pnpm) : 
https://github.com/changesets/changesets/blob/34933f041ab87d0fd1dc0250b894e011cf12da00/packages/cli/src/commands/publish/npm-utils.ts#L39

This forces users to maintain yarn and npm configurations in order for example to publish to private repos.
What's the reason behind this choice ?

## Proposed solution
Support yarn for publishing.

Thanks in advance.

### LINKED ISSUE #1411: [CLI]: yarn changeset publish is not attempting to publish with yarn
## Problem

I'm working on a yarn monorepo in which the packages must be published to a private registry. Initially when running `yarn changeset publish`, I noticed that it was ignoring the `npmPublishRegistry` field in my .`yarnrc.yml`. I got around the issue by adding the registry to `publishConfig` property of all the `package.json`s, but after successfully publishing the packages, I ran into major issue. Yarn replaces any references to the `workspace` protocol within the `package.json` with the correct version, but since the CLI seems to always publish with npm, this replacement never ocurrs, and thus a broken `package.json` is published, which cannot be installed with npm.

I haven't done a deep dive into the source code, but unless I'm misunderstanding something [getPublishTool](https://github.com/changesets/changesets/blob/main/packages/cli/src/commands/publish/npm-utils.ts#L48) only returns either `npm` or `pnpm`.

## Proposed solution

`getPublishTool` should also support yarn

### LINKED ISSUE #1454: Publishing with yarn doesn't work correctly
## Affected Packages

@changesets/cli

## Problem

Publishing with yarn should be done via `yarn npm publish` but `@changesets/cli` is using `npm publish` instead. This means some workspaces features that are yarn specific do not work as expected. For instance, when publishing with `yarn npm publish` yarn replaces inner packages with version `workspace:^` with the correct version. While when publishing with `npm publish` npm ignores packages with version `workspace:^`. This results in incorrectly published packages because when installing it package managers won't be able to resolve `workspace:^`.

I believe the above is also the reason why `.yarnrc.yml` is ignored when publishing via changesets.

This is simple to reproduce:
1. In a monorepo managed by yarn create packages "@example/a" and "@example/b".
2. The package.json of @example/a contains @example/b as a dependency:
```
"dependencies": {
  "@example/b": "workspace:^"
}
```
3. Create a changeset for @example/a.
4. Generate the new versions via `changeset version`.
5. Then publish via `changeset publish`.
6. Now install the new version in a new project and check the package.json in the `node_modules` and also the new info added to `yarn.lock`. You'll see `workspace:^` hasn't been replaced.

Related issues:
* https://github.com/changesets/changesets/issues/1411
* https://github.com/changesets/changesets/issues/650

## Proposed solution

The proposed solution is to use `yarn npm publish` instead of `npm publish`. I tested this out in a private monorepo and it works. I also opened a PR for this change [here](https://github.com/changesets/changesets/pull/1449) and would like help getting it reviewed and approved.

This issue is pretty urgent for us at my company. Any guidance on how to move this forward would be very much appreciated.

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/changesets/changesets/issues/432
fixes https://github.com/changesets/changesets/issues/1454
fixes https://github.com/changesets/changesets/issues/598
fixes https://github.com/changesets/changesets/issues/813
fixes https://github.com/changesets/changesets/issues/1411

### CODE REVIEW

**Comment by kwangure** [comment]
From what I can tell this does not support Vite Plus via `vp pm publish`. For example, I have Vite+ installed globally, but pnpm is not globally installed. Vite manages its own `pnpm` copies to do `vp install` etc. based on the package manager versions different projects specify. So even though, there's a pnpm-workspace.yaml in my project, I'd like changesets to use `vp` and VIte+'s package manager resolution algo. I imagine this'll be increasingly popular.

**Comment by Andarist** [comment]
This is definitely outside the scope of this PR. Please open a new feature request so we could properly track this.

File: packages/cli/src/commands/__tests__/e2e-utils.ts [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+    });
+    return cwd;
+  };
+}
+
+export const pmCases = [
+  {
+    name: "npm 10",
+    bins: { npm: "npm-10" },
+    gitdir: createNpmGitdir("npm@10.9.8"),
```
Comment (bluwy) [inline packages/cli/src/commands/__tests__/e2e-utils.ts]: Could these versions (e.g. the `10.9.8`) derive from the actual installed version? In case `npm-10` has a new patch release and we update in the package.json, then this seems to diverge.

**Review by bluwy — COMMENTED** [review]
I can't say I understood every edge case and pm-specific handling, but I trust that this is handled well with the e2e tests 😄 And the logic seems good for me. 

Also, we need changesets here.

File: packages/cli/src/commands/__tests__/e2e-utils.ts (hunk 1 above)
Comment (Andarist) [reply packages/cli/src/commands/__tests__/e2e-utils.ts]: yep, I wanted to do that at some point - just focused on other things to get this review-ready. I'll make this change in a moment

Comment (Andarist) [reply packages/cli/src/commands/__tests__/e2e-utils.ts]: done

File: packages/cli/src/commands/publish-plan/getPublishPlan.ts [hunk 2]
```diff
@@ -117,6 +121,9 @@ export async function getUnpublishedPackages(
           if (
             preState != null &&
             response.pkgInfo.versions &&
+            // non-npm registries often don't auto-assign latest and when using those we don't have to care about only-pre case
+            // when the latest tag is not auto-assigned we can simply use the configured pre tag
+            response.pkgInfo["dist-tags"].latest &&
```
Comment (bluwy) [inline packages/cli/src/commands/publish-plan/getPublishPlan.ts:126]: Does this mean we support non-latest publishing? Should we mention in the changeset?

**Review by bluwy — APPROVED** [review]

File: packages/cli/src/commands/publish-plan/getPublishPlan.ts (hunk 2 above)
Comment (Andarist) [reply packages/cli/src/commands/publish-plan/getPublishPlan.ts:126]: I'd like to revisit this behavior - it feels kinda surprising that prerelease packages can be published with the latest tag in prerelease mode (and not with the configured prerelease tag) and it all depends on the registry's state that the user isn't usually quite aware of. 

I added the changeset for this change.

**Comment by jperezrealini** [comment]
Will these changes help solve this issue? https://github.com/changesets/changesets/issues/1468

**Comment by Andarist** [comment]
@jperezrealini not yet - but we'll work on Bun support soon. This PR and https://github.com/changesets/changesets/pull/2155 prepare the groundwork for that
